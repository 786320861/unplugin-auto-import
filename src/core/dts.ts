import { ImportsFlatMap } from '../types'

export function generateDeclration(imports: ImportsFlatMap) {
  const body = Object
    .entries(imports)
    .map(([name, info]) => `  const ${name}: typeof import('${info.module}')['${info.from || name}']`)
    .join('\n')
  return `// Generated by 'unplugin-auto-import'\ndeclare global {\n${body}\n}\nexport {}\n`
}
