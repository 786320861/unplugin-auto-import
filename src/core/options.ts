import { resolve } from 'path'
import { toArray } from '@antfu/utils'
import { createFilter } from '@rollup/pluginutils'
import { presets } from '../presets'
import { ImportInfo, ImportsFlatMap, Options, ResolvedOptions } from '../types'

export function resolveOptions(options: Options = {}): ResolvedOptions {
  const imports = flattenImportsMap(options.imports)
  const resolved: ResolvedOptions = {
    ...options,
    dts: options.dts === false
      ? false
      : resolve(options.dts || 'auto-imports.d.ts'),
    imports,
    matchRE: new RegExp(`\\b(${Object.keys(imports).join('|')})\\b`, 'g'),
    idFilter: createFilter(
      options.include || [/\.[jt]sx?$/, /\.vue\??/],
      options.exclude || [/node_modules/, /\.git/],
    ),
  }

  return resolved
}

export function flattenImportsMap(map: Options['imports']): ImportsFlatMap {
  const flat: ImportsFlatMap = {}
  toArray(map).forEach((definition) => {
    if (typeof definition === 'string') {
      if (!presets[definition])
        throw new Error(`[global-import] preset ${definition} not found`)
      definition = presets[definition]
    }

    for (const mod of Object.keys(definition)) {
      for (const id of definition[mod]) {
        const meta = {
          module: mod,
        } as ImportInfo
        if (Array.isArray(id)) {
          meta.name = id[1]
          meta.from = id[0]
        }
        else {
          meta.name = id
        }

        if (flat[meta.name])
          throw new Error(`[global-import] identifier ${meta.name} already defined with ${flat[meta.name].module}`)

        flat[meta.name] = meta
      }
    }
  })

  return flat
}
