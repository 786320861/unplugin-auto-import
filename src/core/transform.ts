import MagicString from 'magic-string'
import { ImportInfo, TransformOptions } from '../types'

const excludeRegex = [
  // imported from other module
  /\bimport\s*\{([\s\S]*?)\}\s*from\b/g,
  // defined as function
  /\bfunction\s*([\s\S]+?)\s*\(/g,
  // defined as local variable
  /\b(?:const|let|var)\s*([\w\d_$]+?)\b/g,
]

export function transform(code: string, id: string, { matchRE, imports }: TransformOptions) {
  const matched = new Set(Array.from(code.matchAll(matchRE)).map(i => i[1]))

  // remove those already defined
  for (const regex of excludeRegex) {
    Array.from(code.matchAll(regex))
      .flatMap(i => i[1]?.split(',') || [])
      .forEach(i => matched.delete(i.trim()))
  }

  const modules: Record<string, ImportInfo[]> = {}

  // group by module name
  Array.from(matched).forEach((name) => {
    const moduleName = imports[name]?.module
    if (!moduleName)
      return
    if (!modules[moduleName])
      modules[moduleName] = []
    modules[moduleName].push(imports[name])
  })

  // stringify import
  const importStatements = Object.entries(modules)
    .map(([moduleName, names]) => {
      const imports = names
        .map(({ name, from }) => from ? `${from} as ${name}` : name)
        .join(', ')
      return `import { ${imports} } from '${moduleName}';`
    })
    .join('')

  const s = new MagicString(code)
  s.prependLeft(0, importStatements)

  return {
    code: s.toString(),
    get map() {
      return s.generateMap()
    },
  }
}
