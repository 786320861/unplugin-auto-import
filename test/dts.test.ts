import { generateDeclration } from '../src/core/dts'
import { resolveOptions } from '../src/core/options'

it('dts', () => {
  const options = resolveOptions({
    imports: [
      'vue-demi',
      'react',
      {
        custom: [
          'customNamed',
          ['default', 'customDefault'],
        ],
      },
    ],
  })

  expect(generateDeclration(options.imports)).toMatchSnapshot()
})
