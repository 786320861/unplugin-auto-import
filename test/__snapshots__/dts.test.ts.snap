// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dts 1`] = `
"// Generated by 'unplugin-auto-import'
declare global {
  const onActivated: typeof import('vue-demi')['onActivated']
  const onBeforeMount: typeof import('vue-demi')['onBeforeMount']
  const onBeforeUnmount: typeof import('vue-demi')['onBeforeUnmount']
  const onBeforeUpdate: typeof import('vue-demi')['onBeforeUpdate']
  const onMounted: typeof import('vue-demi')['onMounted']
  const onUnmounted: typeof import('vue-demi')['onUnmounted']
  const onUpdated: typeof import('vue-demi')['onUpdated']
  const computed: typeof import('vue-demi')['computed']
  const customRef: typeof import('vue-demi')['customRef']
  const isReadonly: typeof import('vue-demi')['isReadonly']
  const isRef: typeof import('vue-demi')['isRef']
  const markRaw: typeof import('vue-demi')['markRaw']
  const reactive: typeof import('vue-demi')['reactive']
  const readonly: typeof import('vue-demi')['readonly']
  const ref: typeof import('vue-demi')['ref']
  const shallowReactive: typeof import('vue-demi')['shallowReactive']
  const shallowReadonly: typeof import('vue-demi')['shallowReadonly']
  const shallowRef: typeof import('vue-demi')['shallowRef']
  const toRaw: typeof import('vue-demi')['toRaw']
  const toRef: typeof import('vue-demi')['toRef']
  const toRefs: typeof import('vue-demi')['toRefs']
  const unref: typeof import('vue-demi')['unref']
  const watch: typeof import('vue-demi')['watch']
  const watchEffect: typeof import('vue-demi')['watchEffect']
  const defineComponent: typeof import('vue-demi')['defineComponent']
  const getCurrentInstance: typeof import('vue-demi')['getCurrentInstance']
  const h: typeof import('vue-demi')['h']
  const inject: typeof import('vue-demi')['inject']
  const nextTick: typeof import('vue-demi')['nextTick']
  const provide: typeof import('vue-demi')['provide']
  const useCssModule: typeof import('vue-demi')['useCssModule']
  const createApp: typeof import('vue-demi')['createApp']
  const useState: typeof import('react')['useState']
  const useCallback: typeof import('react')['useCallback']
  const useMemo: typeof import('react')['useMemo']
  const useEffect: typeof import('react')['useEffect']
  const useRef: typeof import('react')['useRef']
  const useContext: typeof import('react')['useContext']
  const useReducer: typeof import('react')['useReducer']
  const customNamed: typeof import('custom')['customNamed']
  const customDefault: typeof import('custom')['default']
}
export {}
"
`;
